packages/flutter/lib/src/scheduler/scheduler.dart

```diff
@@ -0,0 +1,149 @@
+// Copyright 2015 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:async';
+
+import 'package:collection/priority_queue.dart';
+import 'package:flutter/animation.dart' as animation show scheduler;
+
+typedef Task();
+
+/// An entry in the scheduler's priority queue.
+///
+/// Combines the task and its priority.
+class _SchedulerEntry {
+  final Task task;
+  final int priority;
+
+  _SchedulerEntry(this.task, this.priority);
+}
+
+class Priority {
+  static const kIdle = const Priority._(0);
+  static const kAnimation = const Priority._(100000);
+  static const kTouch = const Priority._(200000);
+
+  /// Relative priorities are clamped by this offset.
+  ///
+  /// It is still possible to have priorities that are offset by more than this
+  /// amount by repeatedly taking relative offsets, but that's generally
+  /// discouraged.
+  static const kMaxOffset = 10000;
+
+  final int _value;
+
+  int get value => _value;
+
+  const Priority._(this._value);
+
+  /// Returns a priority relative to this priority.
+  ///
+  /// A positive [offset] indicates a higher priority.
+  ///
+  /// The parameter [offset] is clamped to +/-[kMaxOffset].
+  Priority operator +(int offset) {
+    if (offset.abs() > kMaxOffset) {
+      // Clamp the input offset.
+      offset = kMaxOffset * offset.sign;
+    }
+    return new Priority._(_value + offset);
+  }
+
+  /// Returns a priority relative to this priority.
+  ///
+  /// A positive offset indicates a lower priority.
+  ///
+  /// The parameter [offset] is clamped to +/-[kMaxOffset].
+  Priority operator -(int offset) => this + (-offset);
+}
+
+/// Scheduler running tasks with specific priorities.
+///
+/// Combines the task's priority with remaining time in a frame to decide when
+/// the task should be run.
+///
+/// Tasks always run in the idle time after a frame has been committed.
+class TaskScheduler {
+  final PriorityQueue _queue = new HeapPriorityQueue<_SchedulerEntry>(
+    (_SchedulerEntry e1, _SchedulerEntry e2) {
+      // Note that we inverse the priority.
+      return -e1.priority.compareTo(e2.priority);
+    });
+
+  SchedulingStrategy schedulingStrategy = new SchedulingStrategy();
+
+  /// Wether this scheduler already requested to be woken up as soon as
+  /// possible.
+  bool _wakingNow = false;
+  /// Wether this scheduler already requested to be woken up in the next frame.
+  bool _wakingNextFrame = false;
+
+  TaskScheduler._();
+
+  /// Schedules the given [task] with the given [priority].
+  // TODO(floitsch): provide some means to indicate how long things are going
+  // to take?
+  // TODO(floitsch): guidance on how to increase priority over time?
```

Comment:
I would either remove these todos (if we have no use case, then there's no need to add it; when we have a use case, we can add it even without the todo), or rephrase it as "other features that could be added include...". We try to avoid checking in questions.